### PR TITLE
Feature/#26 modmail command

### LINKED ISSUE #26: Modmail command
A command which lets users write a message to all the mods. This could be achieved by posting them in a specified channel in the hidden category for moderators.

### PR DESCRIPTION
**Pull request by PKlempe**
Closes #26

### CODE REVIEW

File: bot/persistence/resources/init_db.sql [hunk 1]
```diff
@@ -2,3 +2,8 @@ CREATE TABLE IF NOT EXISTS configs(
     config_key varchar PRIMARY KEY,
     val varchar NOT NULL
 );
+
+CREATE TABLE IF NOT EXISTS Modmail(
```
Comment (SebastiantheAstorian) [inline bot/persistence/resources/init_db.sql:6]: We should be consistent with column and table naming strategies.
I prefer lowercase column names (single words) and lower snake_case table names. But feel free to convince me otherwise. Just make it consistent

File: bot/moderation/moderation.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+        Args:
+            ctx (Context): The context in which the command was called.
+        """
+        msg_content = ctx.message.content
+        msg_attachments = ctx.message.attachments
+        await ctx.message.delete()
+
+        ch_modmail = ctx.guild.get_channel(constants.CHANNEL_ID_MODMAIL)
+        msg_content = msg_content[len(ctx.prefix + ctx.command.name):]
```
[the remark this answers is not in the document]
Comment (PKlempe) [reply bot/moderation/moderation.py:39]: Remove the leading prefix + command from the message.
E.g.: "!modmail This is a Test." => "This is a Test."

File: bot/moderation/moderation.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+        embed = discord.Embed(title="Status: Offen",
```
[the remark this answers is not in the document]
Comment (PKlempe) [reply bot/moderation/moderation.py:41]: We could do that but I'm unsure what the best approach would be for that.

File: bot/moderation/moderation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (SebastiantheAstorian) [reply bot/moderation/moderation.py:39]: Maybe add 1 to it to also remove the leading space after the command

and make a comment what this does pls.

File: bot/moderation/moderation.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        await msg_modmail.add_reaction(constants.EMOJI_MODMAIL_ASSIGN)
+
+        embed_confirmation = embed.to_dict()
+        embed_confirmation["title"] = "Deine Nachricht:"
+        embed_confirmation["color"] = constants.EMBED_COLOR_INFO
+        embed_confirmation = discord.Embed.from_dict(embed_confirmation)
+        await ctx.author.send("Deine Nachricht wurde erfolgreich an die Moderatoren weitergeleitet!\n"
+                              "__Hier deine Bestätigung:__", embed=embed_confirmation)
+
+    @commands.Cog.listener(name='on_raw_reaction_add')
```
[the remark this answers is not in the document]
Comment (PKlempe) [reply bot/moderation/moderation.py:59]: `on_reaction_add()` only works if the message is in the internal message cache of Discord which I think contains the last 5000 messages. 
The advantage of `on_raw_reaction_add()` is that it works regardless of the state of the internal cache.

File: bot/moderation/moderation.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    async def change_modmail_status_add(self, payload):
+        """Event listener which changes the status and the embed of a modmail message if a specific reaction has been
+        added by a user.
+
+        Args:
+            payload (discord.RawReactionActionEvent): The payload for the triggered event.
+        """
+        if not payload.member.bot and payload.channel_id == constants.CHANNEL_ID_MODMAIL:
+            modmail = await self.bot.get_channel(payload.channel_id).fetch_message(payload.message_id)
```
[the remark this answers is not in the document]
Comment (PKlempe) [reply bot/moderation/moderation.py:68]: It will probably throw an exception but since this is an event which triggers when someone **reacts to a message** this should never happen. The only way something could go wrong is if someone manually deletes an entry in our DB or if we implement a bug which does something similar.

So, idk. What do you think?

File: bot/moderation/moderation.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+            await modmail.edit(embed=discord.Embed.from_dict(dict_embed))
+
+    @commands.Cog.listener(name='on_raw_reaction_remove')
+    async def change_modmail_status_remove(self, payload):
+        """Event listener which changes the status and the embed of a modmail message if a specific reaction has been
+        removed by a user.
+
+        Args:
+            payload (discord.RawReactionActionEvent): The payload for the triggered event.
```
[the remark this answers is not in the document]
Comment (PKlempe) [reply bot/moderation/moderation.py:90]: Nope, because the bot will never remove his own reactions.

File: bot/moderation/moderation.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+            dict_embed = modmail.embeds[0].to_dict()
+            dict_embed["title"] = "Status: Offen"
+
+            if payload.emoji.name == constants.EMOJI_MODMAIL_DONE and curr_status != ModmailStatus.OPEN:
+                self._db_connector.change_modmail_status(modmail.id, ModmailStatus.OPEN)
+                dict_embed["color"] = constants.EMBED_COLOR_MODMAIL_OPEN
+                await modmail.add_reaction(constants.EMOJI_MODMAIL_ASSIGN)
+            elif payload.emoji.name == constants.EMOJI_MODMAIL_ASSIGN and curr_status != ModmailStatus.OPEN:
+                self._db_connector.change_modmail_status(modmail.id, ModmailStatus.OPEN)
+                dict_embed["color"] = constants.EMBED_COLOR_MODMAIL_OPEN
```
[the remark this answers is not in the document]
Comment (PKlempe) [reply bot/moderation/moderation.py:104]: + 📝 -> Status = In Progress
-  📝 -> Status = Open
+ ✅ -> Status = Closed
-  ✅ -> Status = Open

Keep in mind that setting the status to _closed_ is always going to remove/clear the paper reaction.

File: bot/moderation/modmail_status.py [hunk 8]
```diff
@@ -0,0 +1,10 @@
+"""Module containing a single Enum class for Modmails."""
```
[the remark this answers is not in the document]
Comment (PKlempe) [reply bot/moderation/modmail_status.py:1]: Yeah I know but I kinda like to have an own file for Enums. This way I can always find them very easily but I think thats just my personal taste.

File: bot/persistence/database_connector.py [hunk 9]
```diff
@@ -1,37 +1,34 @@
...
         """
+        if db_file is None:
+            raise Error("Database filepath and/or filename hasn't been set.")
+
         self._db_file = db_file
         with DatabaseManager(self._db_file) as db_manager:
             if init_script is not None:
-                queries = self.parse_sql_file(init_script)
-                for query in queries:
+                queries_ = self.parse_sql_file(init_script)
```
[the remark this answers is not in the document]
Comment (PKlempe) [reply bot/persistence/database_connector.py:30]: Because `queries` was overshadowing our `queries.py`. This didn't really affect our code but IntelliJ was whining about it. That's why I chose to add a trailing underscore because the name is fitting and I wanted to avoid this naming conflict.

File: bot/persistence/database_connector.py [hunk 10]
```diff
@@ -49,11 +46,8 @@ def add_config_property(self, key: str, val: str):
                 is needed to operate on the db. This also ensures that the database exists and works, contains the table
                 prior to calling this method.
         """
-        if self._db_file is None:
-            raise Error(
-                "Method __init__(filename, init_script=None) has not yet been called. Database filename is not set.")
         with DatabaseManager(self._db_file) as db_manager:
-            db_manager.execute(INSERT_PROPERTY_QUERY, (key, val))
+            db_manager.execute(queries.QUERY_INSERT_PROPERTY, (key, val))
```
[the remark this answers is not in the document]
Comment (PKlempe) [reply bot/persistence/database_connector.py:50]: Because the more features we implement, the more queries we would have to manually import which I thought would be kinda tideous. That's why I imported the whole module.

File: bot/persistence/database_connector.py [hunk 11]
```diff
@@ -63,19 +57,54 @@ def get_property(self, key: str):
...
+
+        Args:
+            msg_id (int): The message id of the modmail.
+
+        Returns:
+            Optional[ModmailStatus]: The current status of the modmail.
+        """
+        with DatabaseManager(self._db_file) as db_manager:
+            result = db_manager.execute(queries.QUERY_GET_MODMAIL_STATUS, (msg_id,))
+            db_manager.commit()
```
[the remark this answers is not in the document]
Comment (PKlempe) [reply bot/persistence/database_connector.py:90]: You are right. That's a copy & paste error.

File: bot/persistence/queries.py [hunk 12]
```diff
@@ -1,4 +1,10 @@
 """This module contains template string constants which will be used for database queries and commands."""
 
-INSERT_PROPERTY_QUERY = "INSERT OR REPLACE INTO configs VALUES(?, ?);"
-GET_PROPERTY_QUERY = "SELECT val FROM configs WHERE config_key = ?"
+QUERY_INSERT_PROPERTY = "INSERT OR REPLACE INTO configs VALUES(?, ?);"
```
[the remark this answers is not in the document]
Comment (PKlempe) [reply bot/persistence/queries.py:3]: 1. I was planning to do so but forgot about it. Thanks for that!
2. I think we should simply remove the word `QUERY` completely.

File: bot/persistence/queries.py [hunk 13]
```diff
@@ -1,4 +1,10 @@
...
+QUERY_GET_PROPERTY = "SELECT val FROM configs WHERE config_key = ?"
+
+# Modmail
+QUERY_INSERT_MODMAIL = "INSERT INTO MODMAIL (ID) VALUES (?)"
+QUERY_CHANGE_MODMAIL_STATUS = "UPDATE Modmail SET StatusID = ? WHERE ID = ?"
+QUERY_GET_MODMAIL_STATUS = "SELECT StatusID FROM Modmail WHERE ID = ?"
+QUERY_GET_MODMAIL_OPEN = "SELECT ID FROM Modmail WHERE StatusID = ?"
```
[the remark this answers is not in the document]
Comment (PKlempe) [reply bot/persistence/queries.py:10]: I was trying to make them look consistent and not grammatically correct. :(
Any suggestions for a better naming scheme?

File: bot/persistence/resources/init_db.sql [hunk 14]
```diff
@@ -2,3 +2,8 @@ CREATE TABLE IF NOT EXISTS configs(
...
+    ID INTEGER NOT NULL PRIMARY KEY,
```
[the remark this answers is not in the document]
Comment (PKlempe) [reply bot/persistence/resources/init_db.sql:7]: That's not fully correct. Due to a bug a Primary Key in SQLite is implicitly `Not Null` only if its a `Integer Primary Key` otherwise it's theoratically nullable. 🤡 
https://www.sqlitetutorial.net/sqlite-not-null-constraint/

Discord uses Twitters "Snowflake" format for unique IDs which is generally an unsigned 64Bit Integer. However in the case of Python it doesn't matter how big a value is because in Python, the value of an integer is not restricted by the number of bits and can expand to the limit of the available memory.
https://stackoverflow.com/a/7604981/11889326
**TL;DR: Integers are unbounded in Python.**

File: bot/persistence/resources/init_db.sql (hunk 1 above)
Comment (PKlempe) [reply bot/persistence/resources/init_db.sql:6]: Singular and PascalCase for **_everything_**.

File: bot/moderation/moderation.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (SebastiantheAstorian) [reply bot/moderation/moderation.py:68]: Or if someone creates the message by hand right?

File: bot/moderation/moderation.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (SebastiantheAstorian) [reply bot/moderation/moderation.py:90]: What does clear reactions method do then?

File: bot/moderation/moderation.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (SebastiantheAstorian) [reply bot/moderation/moderation.py:104]: So you cannot directly go from closed to in progress, but only via open?
I think I got it now.

File: bot/moderation/modmail_status.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (SebastiantheAstorian) [reply bot/moderation/modmail_status.py:1]: Nothing wrong with that. I just assumed it would fit to the topic of the cog

File: bot/persistence/database_connector.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (SebastiantheAstorian) [reply bot/persistence/database_connector.py:50]: I personally prefer explicit Imports, especially since it makes the code where the constants are used shorter and easier to read.

File: bot/moderation/moderation.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (PKlempe) [reply bot/moderation/moderation.py:90]: **Remove reaction != Clear reactions**
Clear removes the reactions from all user who have reacted with a specific emoji. That's a different kind of event for Discord.

File: bot/persistence/resources/init_db.sql (hunk 14 above)
[the remark this answers is not in the document]
Comment (SebastiantheAstorian) [reply bot/persistence/resources/init_db.sql:7]: Oh didn't know that

But are they unbounded in sqllite?

File: bot/persistence/resources/init_db.sql (hunk 1 above)
Comment (SebastiantheAstorian) [reply bot/persistence/resources/init_db.sql:6]: Your strong and uncountable argumentation chain convince me :P

File: bot/moderation/moderation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (PKlempe) [reply bot/moderation/moderation.py:39]: > and make a comment what this does pls.
Or you could try to `git gud`... ᴵᵐ ᵍᵒⁿⁿᵃ ᵃᵈᵈ ᵗʰᶦˢ ᶜᵒᵐᵐᵉⁿᵗ ˡᵃᵗᵉʳ.

File: bot/moderation/moderation.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (SebastiantheAstorian) [reply bot/moderation/moderation.py:90]: And it does not trigger this event?

File: bot/moderation/moderation.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (PKlempe) [reply bot/moderation/moderation.py:68]: Yeah, then too.

File: bot/persistence/resources/init_db.sql (hunk 14 above)
[the remark this answers is not in the document]
Comment (PKlempe) [reply bot/persistence/resources/init_db.sql:7]: No... SQLite can only handle signed 64bit Integers. I'm storing the IDs now as `Text `instead of `Integer`.

File: bot/moderation/moderation.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (PKlempe) [reply bot/moderation/moderation.py:90]: No.
